contracts/test/integration/ExcubiaeSemaphorePaymaster.test.ts

```diff
@@ -0,0 +1,247 @@
+import { ethers } from "hardhat";
+import { Signer } from "ethers";
+import {
+  SimpleAccount,
+  ExcubiaeSemaphorePaymaster,
+} from "../../typechain";
+import { expect } from "chai";
+import { Identity, Group } from "@semaphore-protocol/core"
+import { setupProviders, setupSimpleAccount, prepareTransferCallData, prepareUserOp, TestContext, setupExcubiaeSemaphorePaymasterContracts, generateExcubiaePaymasterData } from "../utils/testUtils";
+import sendUserOpAndWait from "../utils/userOpUtils";
+
+const ENABLE_LOGS = false; // Toggle this to enable/disable logging
+const log = (...args: any[]) => ENABLE_LOGS && console.log(...args);
+
+describe("ExcubiaePaymasterTest", () => {
+  let context: TestContext;
+  let owner: Signer;
+  let recipientAddress: string;
+  let simpleAccount: SimpleAccount;
+  let excubiaeSemaphorePaymaster: ExcubiaeSemaphorePaymaster;
+  let group: Group;
+  let validGroupId: number;
+  let invalidGroupId: number;
+  let validEvidence: string;
+  let invalidEvidence: string;
+  let id1: Identity;
+  let id2: Identity;
+  let id3: Identity;
+  const transferAmount = ethers.parseEther("1");
+
+  async function setupIdentities() {
+    const [admin, owner, id1Pk, id2Pk, id3Pk] = await ethers.getSigners();
+
+    id1 = new Identity(id1Pk);
+    id2 = new Identity(id2Pk);
+    id3 = new Identity(id3Pk);
+    group = new Group([id1.commitment, id2.commitment, id3.commitment]);
+    validGroupId = 0;
+    invalidGroupId = 1;
+
+    return { admin, owner };
+  }
+
+  async function assertSendEth(
```

Comment:
Don't need to address this for PR but just noticed this code is duplicated enough times where it could be worth abstracting it into a util. Fyi rather than a required code change

Reply:
Fair point, thanks!